src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,311 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\ResolverInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    const CHECK_INTERVAL = 0.1;
+
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array(
+        Message::TYPE_A    => false,
+        Message::TYPE_AAAA => false,
+    );
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, ResolverInterface $resolver, $uri, $host, $parts)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+        $this->host = $host;
+        $this->parts = $parts;
+    }
+
+    public function connect()
+    {
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            $lookupResolve = function ($type) use ($that, $resolve, $reject) {
+                return function (array $ips) use ($that, $type, $resolve, $reject) {
+                    unset($that->resolverPromises[$type]);
+                    $that->resolved[$type] = true;
+
+                    $that->mixIpsIntoConnectQueue($ips);
+
+                    if ($that->timer instanceof TimerInterface) {
+                        return;
+                    }
+
+                    $that->check($resolve, $reject);
+                };
+            };
+
+            $ipv4Deferred = null;
+            $that->resolverPromises[Message::TYPE_AAAA] = $that->resolve(Message::TYPE_AAAA, $reject)->then($lookupResolve(Message::TYPE_AAAA))->then(function () use (&$ipv4Deferred) {
+                if ($ipv4Deferred instanceof Promise\Deferred) {
+                    $ipv4Deferred->resolve();
+                }
+            });
+            $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $reject)->then(function ($ips) use ($that, &$ipv4Deferred) {
+                if ($that->resolved[Message::TYPE_AAAA] === true) {
+                    return Promise\resolve($ips);
+                }
+
+                /**
+                 * Delay A lookup by 50ms sending out connection to IPv4 addresses when IPv6 records haven't
+                 * resolved yet as per RFC.
+                 *
+                 * @link https://tools.ietf.org/html/rfc8305#section-3
+                 */
+                $ipv4Deferred = new Promise\Deferred();
+                $timer = Promise\Timer\resolve(0.05, $that->loop);
```

Comment:
There's no clean up logic in case the resulting promise is cancelled while this timer is running. I would suggest directly using the loop instead.